### PR TITLE
[ML] Throttle the delete-by-query of expired results

### PR DESCRIPTION
**Pull request by droberts195**
Due to #47003 many clusters will have built up a
large backlog of expired results. On upgrading to
a version where that bug is fixed users could find
that the first ML daily maintenance task deletes
a very large amount of documents.

This change introduces throttling to the
delete-by-query that the ML daily maintenance uses
to delete expired results to limit it to deleting an
average 200 documents per second. (There is no
throttling for state/forecast documents as these
are expected to be lower volume.)

Additionally a rough time limit of 8 hours is applied
to the whole delete expired data action. (This is only
rough as it won't stop part way through a single
operation - it only checks the timeout between
operations.)

Relates #47103

### CODE REVIEW

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM

**Comment by dimitris-athanasiou** [comment]
run elasticsearch-ci/default-distro

**Comment by dimitris-athanasiou** [comment]
run elasticsearch-ci/packaging-sample

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/retention/ExpiredResultsRemover.java [hunk 1]
```diff
@@ -88,6 +88,14 @@ private DeleteByQueryRequest createDBQRequest(Job job, long cutoffEpochMs) {
         DeleteByQueryRequest request = new DeleteByQueryRequest();
         request.setSlices(AbstractBulkByScrollRequest.AUTO_SLICES);
 
+        // Delete the documents gradually.
+        // With DEFAULT_SCROLL_SIZE = 1000 this implies we spread deletion of 1 million documents over 5000 seconds ~= 83 minutes.
+        // And we delete a maximum of 10000 batches per day (= 10 million documents per day if DEFAULT_SCROLL_SIZE = 1000).
+        // If more documents than this have expired then some will be deleted on a subsequent day.
```
[the remark this answers is not in the document]
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/retention/ExpiredResultsRemover.java]: We do not start the next job's deletion before the previous is complete.

[the remark this answers is not in the document]
Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/retention/ExpiredResultsRemover.java]: > We do not start the next job's deletion before the previous is complete

I think the requests per second limit is still valuable, but with a separate scroll per job a 10 million docs limit is pretty useless.  I will instead impose a time limit, say 8 hours, at the outermost level of the expired data removal.

[the remark this answers is not in the document]
Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/retention/ExpiredResultsRemover.java]: > The query does not disable scoring, does it make sense to wrap it as constant_score?

I think sorting by `_doc` disables scoring.  This is also the simplest order to retrieve the documents from Lucene.  So I will add a sort by `_doc` to all our delete-by-query searches related to expired data and it should give both benefits.

**Pull request merged**